mentat/interface.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Callable, Dict, Iterable, Optional
+
+from termcolor import cprint
+
+_interface_instance = None
+
+
+def get_interface() -> MentatInterface:
+    if _interface_instance is None:
+        raise Exception("Mentat interface not initialized")
+    return _interface_instance
+
+
+# convenience function for common interaction
+def output(content: str, color: str):
+    get_interface().interact(content=content, color=color)
+
+
+class InterfaceType(Enum):
+    Terminal = "terminal"
+    VSCode = "vscode"
+
+
+def initialize_mentat_interface(
+    interface_type: InterfaceType,
+    interaction_callback: Optional[Callable[[Dict], str]] = None,
+):
+    global _interface_instance
+
+    # should only be initizalized once
+    assert _interface_instance is None
+
+    # callback provided iff interface_type is VSCode
+    assert (interface_type == InterfaceType.VSCode) == (
+        interaction_callback is not None
+    )
+
+    match interface_type:
+        case InterfaceType.Terminal:
+            _interface_instance = TerminalMentatInterface()
+        case InterfaceType.VSCode:
+            _interface_instance = VSCodeMentatInterface(interaction_callback)
+
+
+class MentatInterface(ABC):
+    @abstractmethod
+    def interact(
+        self,
+        content: Optional[str] = None,
+        color: Optional[str] = None,
+        is_error: Optional[bool] = False,
+        return_user_input: Optional[bool] = False,
+        user_input_options: Optional[Iterable[str]] = None,
+    ) -> str:
+        """Handle both receiving input and sending output using the provided callback."""
+        pass
```

Comment:
Actually instead of "interact" we can have a few different functions, input, output, whatever we want. Then the specific `MentatInterface` implementation (like `VSCodeMentatInterface` can decide if it's best to use a single or multiple callbacks to handle them.

Reply:
Yup I was thinking the same thing. Left thoughts here: https://github.com/biobootloader/mentat/issues/63#issuecomment-1694540114